Doc/c-api/lifecycle.rst

```diff
@@ -0,0 +1,236 @@
+
+
+Cyclic Isolate Destruction
+--------------------------
+
+Listed below are the stages of life of a hypothetical :term:`cyclic isolate`
+that continues to exist after each member object is finalized or cleared.  It
+is a bug if a cyclic isolate progresses through all of these stages; it should
+vanish once all objects are cleared, if not sooner.  A cyclic isolate can
+vanish either because the reference cycle is broken or because the objects are
+no longer isolated due to finalizer resurrection (see
+:c:member:`~PyTypeObject.tp_finalize`).
+
+0. **Reachable** (pre-cyclic isolate): All objects are in their normal,
+   reachable state.  A reference cycle exists, but an external reference means
+   the objects are not yet isolated.
+#. **Unreachable but consistent:** The final reference from outside the group
+   of objects has been removed, causing the objects to become isolated (thus a
+   cyclic isolate is born).  None of the group's objects have been finalized or
+   cleared yet.
+#. **Mix of finalized and not finalized:** Objects in a cyclic isolate are
+   finalized one at a time, which means that there is a period of time when the
+   cyclic isolate is composed of a mix of finalized and non-finalized objects.
+   Finalization order is unspecified, so it can appear random.  A finalized
+   object must behave in a sane manner when non-finalized objects interact with
+   it, and a non-finalized object must be able to tolerate the finalization of
+   an arbitrary subset of its referents.
+#. **All finalized:** All objects in a cyclic isolate are finalized before any
+   of them are cleared.
+#. **Mix of finalized and cleared:** The objects can be cleared serially or
+   concurrently (but with the :term:`GIL` held :ref:`unless disabled
+   <whatsnew313-free-threaded-cpython>`); either way, some will finish before
+   others.  A finalized object must be able to tolerate the clearing of a
+   subset of its referents.  :pep:`442` calls this stage "cyclic trash".
+#. **Leaked:** If a cyclic isolate still exists after all objects in the group
+   have been finalized and cleared, then the objects remain indefinitely
+   uncollectable (see :data:`gc.garbage`).  It is a bug if a cyclic isolate
+   reaches this stage---it means the :c:member:`~PyTypeObject.tp_clear` methods
+   of the participating objects have failed to break the reference cycle as
+   required.
+
+If :c:member:`~PyTypeObject.tp_clear` did not exist, then Python would have no
+way to break a reference cycle.  The destruction of an object in a cyclic
+isolate would result in a dangling pointer, triggering undefined behavior when
+an object referencing the destroyed object is itself destroyed.
+
+:c:member:`~PyTypeObject.tp_finalize` is not needed to safely destroy a cyclic
+isolate, but its existence makes it easier to design types that behave in a
+sane manner when objects are cleared.  Clearing an object might necessarily
```

Comment:
It could also be worth mentioning that it's completely safe to execute arbitary Python code in `tp_finalize`. In a lot of cases, it's not necessarily used for destroying the object, but just for doing something upon deletion (e.g. closing a file).